src/librustc/mir/interpret/error.rs

```diff
@@ -137,11 +137,14 @@ impl<'tcx> ConstEvalErr<'tcx> {
         message: &str,
         lint_root: Option<hir::HirId>,
     ) -> Result<DiagnosticBuilder<'tcx>, ErrorHandled> {
+        use InvalidProgramInfo::*;
         match self.error {
-            InterpError::Layout(LayoutError::Unknown(_)) |
-            InterpError::TooGeneric => return Err(ErrorHandled::TooGeneric),
-            InterpError::Layout(LayoutError::SizeOverflow(_)) |
-            InterpError::TypeckError => return Err(ErrorHandled::Reported),
+            InterpError::InvalidProgram(Layout(LayoutError::Unknown(_))) |
+            InterpError::InvalidProgram(TooGeneric) =>
+                return Err(ErrorHandled::TooGeneric),
+            InterpError::InvalidProgram(Layout(LayoutError::SizeOverflow(_))) |
+            InterpError::InvalidProgram(TypeckError) =>
+                return Err(ErrorHandled::Reported),
```

Comment:
This actually exhaustively covers the cases of `InterpError::InvalidProgram`, right? Would be nice if the code could express that, something like
```rust
if let InterpError::InvalidProgram(invalid_err) = self.error {
  match invalid_err {
    // exhaustive match goes here
  }
}
```
```

Reply:
Oh, actually it does not exhaustively cover them. There is also `ReferencedConstant`. So never mind.